### ISSUE TITLE
Using the optimize tag throws a 500 Setting unknown property

### ISSUE THREAD

**Issue by Pautomagi** [barrelstrength/craft-sprout-seo#113]
### Description

The beta 7 version throws a `Setting unknown property:Barrelstrength\sproutseo\models\Metadata::id`

error when  adding the `{% sproutseo 'optimize' %}` tag.

### Steps to reproduce

1. Install beta 7 version of the plugin
2. Add optimize tag

### Additional info

- Craft version: 3.0.24
- PHP version: 7.1.7
- Database driver & version: Mysql 5.7.18
- Plugins & versions: Sprout SEO 4.0.0-beta.7

<details> 
  <summary>Stacktrace</summary>
```
1. in /var/www/html/vendor/yiisoft/yii2/base/Component.php at line 209
200201202203204205206207208209210211212213214215216217218                                    $behavior->$name = $value;
                return;
            }
        }
 
        if (method_exists($this, 'get' . $name)) {
            throw new InvalidCallException('Setting read-only property: ' . get_class($this) . '::' . $name);
        }
 
        throw new UnknownPropertyException('Setting unknown property: ' . get_class($this) . '::' . $name);
    }
 
    /**
     * Checks if a property is set, i.e. defined and not null.
     *
     * This method will check in the following order and act accordingly:
     *
     *  - a property defined by a setter: return whether the property is set
     *  - a property of a behavior: return whether the property is set
                
2. in /var/www/html/vendor/yiisoft/yii2/BaseYii.php at line 546 – yii\base\Component::__set('id', null)
3. in /var/www/html/vendor/yiisoft/yii2/base/BaseObject.php at line 107 – yii\BaseYii::configure(barrelstrength\sproutseo\models\Metadata, ['id' => null, 'isNew' => false, 'elementId' => null, 'default' => null, ...])
101102103104105106107108109110111112113                         *
     * @param array $config name-value pairs that will be used to initialize the object properties
     */
    public function __construct($config = [])
    {
        if (!empty($config)) {
            Yii::configure($this, $config);
        }
        $this->init();
    }
 
    /**
     * Initializes the object.
                
4. in /var/www/html/vendor/barrelstrength/sprout-seo/src/services/Optimize.php at line 223 – yii\base\BaseObject::__construct(['id' => null, 'isNew' => false, 'elementId' => null, 'default' => null, ...])
5. in /var/www/html/vendor/barrelstrength/sprout-seo/src/services/Optimize.php at line 121 – barrelstrength\sproutseo\services\Optimize::getPrioritizedMetadataModel(craft\elements\Entry, craft\models\Site)
115116117118119120121122123124125126127                            /**
         * @var Settings $settings
         */
        $settings = Craft::$app->plugins->getPlugin('sprout-seo')->getSettings();
 
        $this->globals = SproutSeo::$app->globalMetadata->getGlobalMetadata($site);
        $this->prioritizedMetadataModel = $this->getPrioritizedMetadataModel($element, $site);
 
        $output = null;
 
        $metadata = [
            'globals' => $this->globals,
            'meta' => $this->prioritizedMetadataModel->getMetaTagData(),
                
6. in /var/www/html/vendor/barrelstrength/sprout-seo/src/services/Optimize.php at line 99 – barrelstrength\sproutseo\services\Optimize::getMetadata(craft\elements\Entry, craft\models\Site)
93949596979899100101102103104105                            }
 
        if ($element !== null) {
            $this->elementMetadata = SproutSeo::$app->elementMetadata->getElementMetadata($element);
        }
 
        return $this->getMetadata($element, $site);
    }
 
    /**
     * @param      $element
     * @param      $site
     * @param bool $render
                
7. in /var/www/html/storage/runtime/compiled_templates/13/134095be8ed405db4e39eaa5766ca195707a5c10a9cb083c1fcd5544ea2258fb.php at line 48 – barrelstrength\sproutseo\services\Optimize::getMetadataViaContext(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...])
8. in /var/www/html/vendor/twig/twig/lib/Twig/Template.php at line 390 – __TwigTemplate_dbcb15a040d44c82ea260dc2f6d1860dc9e099c90c9bd5326b3ad1e10e8bcd9e::doDisplay('[<span class="string">'entry'</s...', ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'foot' => [__TwigTemplate_dbcb15a040d44c82ea260dc2f6d1860dc9e099c90c9bd5326b3ad1e10e8bcd9e, 'block_foot'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
9. in /var/www/html/vendor/craftcms/cms/src/web/twig/Template.php at line 49 – Twig_Template::displayWithErrorHandling(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'foot' => [__TwigTemplate_dbcb15a040d44c82ea260dc2f6d1860dc9e099c90c9bd5326b3ad1e10e8bcd9e, 'block_foot'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
10. in /var/www/html/vendor/twig/twig/lib/Twig/Template.php at line 367 – craft\web\twig\Template::displayWithErrorHandling(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'foot' => [__TwigTemplate_dbcb15a040d44c82ea260dc2f6d1860dc9e099c90c9bd5326b3ad1e10e8bcd9e, 'block_foot'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
11. in /var/www/html/vendor/craftcms/cms/src/web/twig/Template.php at line 31 – Twig_Template::display(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
12. in /var/www/html/storage/runtime/compiled_templates/1d/1d906d5e4fbce33e41f503427cad4da46d21212d2390ad810b141fac05009005.php at line 143 – craft\web\twig\Template::display(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
13. in /var/www/html/vendor/twig/twig/lib/Twig/Template.php at line 390 – __TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6::doDisplay(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
14. in /var/www/html/vendor/craftcms/cms/src/web/twig/Template.php at line 49 – Twig_Template::displayWithErrorHandling(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
15. in /var/www/html/vendor/twig/twig/lib/Twig/Template.php at line 367 – craft\web\twig\Template::displayWithErrorHandling(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['body' => [__TwigTemplate_a57ba35ee7be85f1b29988c056059c88afd40e3ffe3c531472c34685bc19d5b6, 'block_body'], 'main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
16. in /var/www/html/vendor/craftcms/cms/src/web/twig/Template.php at line 31 – Twig_Template::display(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
17. in /var/www/html/storage/runtime/compiled_templates/18/18344b098900954c441122bc9abd51435d9d9633bd506ba71a1fe776871e6d24.php at line 33 – craft\web\twig\Template::display(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
18. in /var/www/html/vendor/twig/twig/lib/Twig/Template.php at line 390 – __TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6::doDisplay(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
19. in /var/www/html/vendor/craftcms/cms/src/web/twig/Template.php at line 49 – Twig_Template::displayWithErrorHandling(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
20. in /var/www/html/vendor/twig/twig/lib/Twig/Template.php at line 367 – craft\web\twig\Template::displayWithErrorHandling(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry], 'view' => craft\web\View, 'SORT_ASC' => 4, ...], ['main' => [__TwigTemplate_7fc89c7bc7dd6e9c26d88b1ae307b1936fce1284095665b43d42170c19a3fbf6, 'block_main']])
21. in /var/www/html/vendor/craftcms/cms/src/web/twig/Template.php at line 31 – Twig_Template::display(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]], [])
22. in /var/www/html/vendor/twig/twig/lib/Twig/Template.php at line 375 – craft\web\twig\Template::display(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]])
23. in /var/www/html/vendor/twig/twig/lib/Twig/Environment.php at line 289 – Twig_Template::render(['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]])
24. in /var/www/html/vendor/craftcms/cms/src/web/View.php at line 331 – Twig_Environment::render('index', ['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]])
325326327328329330331332333334335336337                            // Render and return
        $renderingTemplate = $this->_renderingTemplate;
        $this->_renderingTemplate = $template;
        Craft::beginProfile($template, __METHOD__);
 
        try {
            $output = $this->getTwig()->render($template, $variables);
        } catch (\RuntimeException $e) {
            if (!YII_DEBUG) {
                // Throw a generic exception instead
                throw new Exception('An error occurred when rendering a template.', 0, $e);
            }
            throw $e;
                
25. in /var/www/html/vendor/craftcms/cms/src/web/View.php at line 378 – craft\web\View::renderTemplate('index', ['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]])
372373374375376377378379380381382383384                            ob_implicit_flush(false);
 
        $isRenderingPageTemplate = $this->_isRenderingPageTemplate;
        $this->_isRenderingPageTemplate = true;
 
        $this->beginPage();
        echo $this->renderTemplate($template, $variables);
        $this->endPage();
 
        $this->_isRenderingPageTemplate = $isRenderingPageTemplate;
 
        $output = ob_get_clean();
 
                
26. in /var/www/html/vendor/craftcms/cms/src/web/Controller.php at line 155 – craft\web\View::renderPageTemplate('index', ['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]])
149150151152153154155156157158159160161                                }
 
            $headers->set('content-type', $mimeType . '; charset=' . $response->charset);
        }
 
        // Render and return the template
        $response->data = $this->getView()->renderPageTemplate($template, $variables);
 
        // Prevent a response formatter from overriding the content-type header
        $response->format = YiiResponse::FORMAT_RAW;
 
        return $response;
    }
                
27. in /var/www/html/vendor/craftcms/cms/src/controllers/TemplatesController.php at line 78 – craft\web\Controller::renderTemplate('index', ['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]])
72737475767778798081828384                     
        // Merge any additional route params
        $routeParams = Craft::$app->getUrlManager()->getRouteParams();
        unset($routeParams['template'], $routeParams['template']);
        $variables = array_merge($variables, $routeParams);
 
        return $this->renderTemplate($template, $variables);
    }
 
    /**
     * Shows the 'offline' template.
     *
     * @return Response
                
28. craft\controllers\TemplatesController::actionRender('index', ['entry' => craft\elements\Entry, 'variables' => ['entry' => craft\elements\Entry]])
29. in /var/www/html/vendor/yiisoft/yii2/base/InlineAction.php at line 57 – call_user_func_array([craft\controllers\TemplatesController, 'actionRender'], ['index', ['entry' => craft\elements\Entry]])
30. in /var/www/html/vendor/yiisoft/yii2/base/Controller.php at line 157 – yii\base\InlineAction::runWithParams(['template' => 'index', 'variables' => ['entry' => craft\elements\Entry]])
31. in /var/www/html/vendor/craftcms/cms/src/web/Controller.php at line 103 – yii\base\Controller::runAction('render', ['template' => 'index', 'variables' => ['entry' => craft\elements\Entry]])
979899100101102103104105106107108109                        /**
     * @inheritdoc
     */
    public function runAction($id, $params = [])
    {
        try {
            return parent::runAction($id, $params);
        } catch (\Throwable $e) {
            if (Craft::$app->getRequest()->getAcceptsJson()) {
                Craft::$app->getErrorHandler()->logException($e);
                $message = $e->getMessage();
                if ($e instanceof ClientException) {
                    $statusCode = $e->getCode();
                
32. in /var/www/html/vendor/yiisoft/yii2/base/Module.php at line 528 – craft\web\Controller::runAction('render', ['template' => 'index', 'variables' => ['entry' => craft\elements\Entry]])
33. in /var/www/html/vendor/craftcms/cms/src/web/Application.php at line 282 – yii\base\Module::runAction('templates/render', ['template' => 'index', 'variables' => ['entry' => craft\elements\Entry]])
276277278279280281282283284285286287288                         * @param string $route
     * @param array $params
     * @return Response|null The result of the action, normalized into a Response object
     */
    public function runAction($route, $params = [])
    {
        $result = parent::runAction($route, $params);
 
        if ($result !== null) {
            if ($result instanceof Response) {
                return $result;
            }
 
                
34. in /var/www/html/vendor/yiisoft/yii2/web/Application.php at line 103 – craft\web\Application::runAction('templates/render', ['template' => 'index', 'variables' => ['entry' => craft\elements\Entry]])
35. in /var/www/html/vendor/craftcms/cms/src/web/Application.php at line 271 – yii\web\Application::handleRequest(craft\web\Request)
265266267268269270271272273274275276277                            // If this is an action request, call the controller
        if (($response = $this->_processActionRequest($request)) !== null) {
            return $response;
        }
 
        // If we're still here, finally let Yii do it's thing.
        return parent::handleRequest($request);
    }
 
    /**
     * @inheritdoc
     * @param string $route
     * @param array $params
                
36. in /var/www/html/vendor/yiisoft/yii2/base/Application.php at line 386 – craft\web\Application::handleRequest(craft\web\Request)
37. in /var/www/html/web/index.php at line 21 – yii\base\Application::run()
15161718192021                        (new Dotenv\Dotenv(CRAFT_BASE_PATH))->load();
}
 
// Load and run Craft
define('CRAFT_ENVIRONMENT', getenv('ENVIRONMENT') ?: 'production');
$app = require CRAFT_VENDOR_PATH.'/craftcms/cms/bootstrap/web.php';
$app->run();

   ```
</details>

**Comment by BenParizek** [comment]
In a discussion on Slack I've confirmed that:

- This situation is happening when updating a site from Craft 2
- The upgrade is being done after updating to the latest version of Sprout SEO on C2
- The only code in the templates is the `{% sproutseo 'optimize' %}` tag (there is no other references to `craft.sproutSeo.meta` or any other tags)

**Comment by BenParizek** [comment]
@Pautomagi We've been unable to recreate this issue after a migration.

Would you be able to send your database and a zip of your templates over to sprout@barrelstrengthdesign.com so we could take a closer look? If you can't share your whole db, if you could share the Sprout SEO specific tables that would also be helpful.

**Comment by BenParizek** [comment]
@Pautomagi This has been resolved in Sprout SEO 4.0.0-beta.9 (update being deployed now), however, you can also resolve the issue after upgrading by resaving your Sprout SEO -> Globals settings.

**Comment by alexagui** [comment]
I ran into this after upgrading our last Craft 2 client to Craft 3.

In case anyone else misses it:  re-Save **`Sprout SEO -> Globals`**
